src/webapis/WebAPIs.ts

```diff
@@ -0,0 +1,209 @@
+                console.debug(`Received a message on ${API_PREFIX}/report/:room_id/:event_id`, request.params);
+                await this.handleReport({ request, response, roomId: request.params.room_id, eventId: request.params.event_id })
+            });
+            console.log(`Configuring ${API_PREFIX}/report/:room_id/:event_id... DONE`);
+        }
+    }
+
+    public stop() {
+        if (this.httpServer) {
+            console.log("Stopping WebAPIs.");
+            this.httpServer.close();
+            this.httpServer = undefined;
+        }
+    }
+
+    /**
+     * Handle a call to the /report API.
+     *
+     * In case of success, respond an empty JSON body.
+     *
+     * @param roomId The room in which the reported event took place. Already extracted from the URL.
+     * @param eventId The event. Already extracted from the URL.
+     * @param request The request. Its body SHOULD hold an object `{reason?: string}`
+     * @param response The response. Used to propagate HTTP success/error.
+     */
+    async handleReport({ roomId, eventId, request, response }: { roomId: string, eventId: string, request: express.Request, response: express.Response }) {
+        // To display any kind of useful information, we need
+        //
+        // 1. The reporter id;
+        // 2. The accused id, to be able to warn/kick/ban them if necessary;
+        // 3. The content of the event **if the room is unencrypted**.
+
+        try {
+            // -- Create a client on behalf of the reporter.
+            // We'll use it to confirm the authenticity of the report.
+            let accessToken;
+
+            // Authentication mechanism 1: Request header.
+            let authorization = request.get('Authorization');
+
+            if (authorization) {
+                [, accessToken] = AUTHORIZATION.exec(authorization)!;
+            } else {
+                // Authentication mechanism 2: Access token as query parameter.
+                accessToken = request.query["access_token"];
+            }
+
+            // Create a client dedicated to this report.
+            //
+            // VERY IMPORTANT NOTES
+            //
+            // We're impersonating the user to get the context of the report.
+            //
+            // For privacy's sake, we MUST ensure that:
+            //
+            // - we DO NOT sync with this client;
+            // - we DO NOT associate a crypto store (e.g. Pantalaimon),
+            //    as this would let us read encrypted messages.
+            let reporterClient = new MatrixClient(config.rawHomeserverUrl, accessToken);
+            reporterClient.start = () => {
+                throw new Error("We MUST NEVER call start on the reporter client");
+            };
+
+            let whoami: any = await reporterClient.doRequest("GET", "/_matrix/client/r0/account/whoami");
+            let reporterId = whoami["user_id"];

```

Comment:
So we have to do this to verify the authenticity of the access token (but also figure out which user this is in the first place)? That's fair enough, but we should probably only use the reporterClient once here, take it out of scope then use mjolnir's client and the admin api for the rest.

Reply:
Good point.  I refactored to use `reporterClient` only for the `whoami` and `/room/.../event/...`. We could remove the need for `/room/.../event/...` and replace this with an Admin API call, but that would restrict this feature to Synapse.

I *think* it's ok to use the `reporterClient` to implement what the Admin API already implements but with a more portable solution. I've added a discussion on the topic in both the code and the README.

Reply:
Ok, i see, that's good :+1: